### PR TITLE
fix: Showing the unique services in the service page

### PR DESCRIPTION
**Pull request by NitishAroraKPMG**
## Description

Fix the list of services displayed in service page.

## Testing instructions

View the operator which is having duplicate services in the "View & Manage fares"

### CODE REVIEW

File: repos/fdbt-site/src/pages/products/services.tsx [hunk 1]
```diff
@@ -187,7 +187,18 @@ export const matchProductsToServices = (
...
     const noc = getAndValidateNoc(ctx);
     const dataSource = !!getSessionAttribute(ctx.req, MULTI_MODAL_ATTRIBUTE) ? 'tnds' : 'bods';
-    const services: MyFaresService[] = await getBodsOrTndsServicesByNoc(noc, dataSource);
+    const allServices: MyFaresService[] = await getBodsOrTndsServicesByNoc(noc, dataSource);
+    const uniqueServiceList: string[] = [];
+    const services: MyFaresService[] = [];
+
+    allServices.forEach((service) => {
+        const service_name = `${service.lineId}#${service.startDate}#${service.endDate}`;
+        if (!uniqueServiceList.includes(service_name)) {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/products/services.tsx:196]: service_name is a python naming standard, we use camelCase in JS/TS.

We need a function which takes an array of services as an argument, and returns an array of services, but removes any duplicates (any services that share the same lineId, startDate and endDate).

Then we need to write unit tests for this function, with different scenarios.

Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/products/services.tsx:196]: Then the usage on this page will be 

` const services: MyFaresService[] = await getBodsOrTndsServicesByNoc(noc, dataSource);`
`const servicesWithNoDuplicates = removeDuplicateServices(services);`

File: repos/fdbt-site/src/data/auroradb.ts [hunk 2]
```diff
@@ -163,6 +163,38 @@ export const getServicesByNocCodeAndDataSource = async (nocCode: string, source:
     }
 };
 
+export const getFaresServicesByNocCodeAndDataSource = async (
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/data/auroradb.ts:166]: Instead of copy and pasting an existing DB query and adding 1 additional piece of information to select, cant we just edit the current DB query (getServicesByNocCodeAndDataSource) and edit the interface ServiceType to include endDate?

Comment (RobCattonKPMG) [reply repos/fdbt-site/src/data/auroradb.ts:166]: Also we presumably need to use the 'convertDateFormat' function on the end date, if its there, like we do with the start date in the query results.

File: repos/fdbt-site/src/utils/index.ts [hunk 3]
```diff
@@ -258,6 +259,25 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
...
-export const fareTypeIsAllowedToAddACap = (fareType: string): boolean => {
+export const fareTypeIsAllowedToAddACap = (fareType: string | undefined): boolean => {
     return !!fareType && ['single', 'return', 'flatFare'].includes(fareType);
 };
+
+export const getUniqueServices = (services: MyFaresService[]): MyFaresService[] => {
+    const uniqueServiceList: string[] = [];
+    const uniqueServices: MyFaresService[] = [];
+
+    if (!!services && services.length > 0) {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:270]: If the function's arguments says it accepts an array of services as an input, we dont need !!services - we know it will be defines. We also dont need to check the length, because if we use a map function, it will return an empty array if it begins with an empty array.

File: repos/fdbt-site/src/pages/service.tsx [hunk 4]
```diff
@@ -24,7 +24,7 @@ const errorId = 'service';
 interface ServiceProps {
     operator: string;
     passengerType: string;
-    services: ServiceType[];
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/service.tsx:27]: I dont think we need to change this type to be MyFaresService - just need to add endDate to ServiceType

File: repos/fdbt-site/src/utils/index.ts [hunk 5]
```diff
@@ -258,6 +259,25 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
     return exportNamePart === exportName;
 };
 
-export const fareTypeIsAllowedToAddACap = (fareType: string): boolean => {
+export const fareTypeIsAllowedToAddACap = (fareType: string | undefined): boolean => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:262]: Why have we changed this to allow undefined to be passed to it?

File: repos/fdbt-site/tests/pages/service.test.tsx [hunk 6]
```diff
@@ -231,32 +230,31 @@ describe('pages', () => {
...
                         {
-                            id: 12,
+                            id: '12',
                             lineName: 'X1',
                             lineId: '3h3vb32ik',
                             startDate: '06/02/2020',
-                            description: 'this bus service is X1',
                             origin: 'Edinburgh',
-                            serviceCode: 'NW_05_BLAC_X1_1',
+                            destination: 'Leeds',
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/pages/service.test.tsx:247]: The fact that all these changed should be setting off alarm bells that we're doing something wrong. We no longer have descripton in the data, and we dont have the serviceCode either. This could (and probably does) break things further down the line.

File: repos/fdbt-site/tests/utils/index.test.ts [hunk 7]
```diff
@@ -249,4 +252,66 @@ describe('index', () => {
...
+                origin: 'Vega de Pas',
+                destination: 'Guiddam',
+                startDate: '05/04/2020',
+                endDate: '25/04/2020',
+            },
+        ];
+
+        it('returns unique services if two services having same lineName, startDate and endDate ', () => {
+            const result = getUniqueServices(services);
+            expect(services.length).toBeLessThanOrEqual(4);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/utils/index.test.ts:313]: We should be checking they return the services we expect, not just checking the length.

Comment (RobCattonKPMG) [reply repos/fdbt-site/tests/utils/index.test.ts:313]: And we also need a test to make sure it doesnt change anything if there are no duplicates.

File: repos/fdbt-site/src/data/auroradb.ts (hunk 2 above)
Comment (NitishAroraKPMG) [reply repos/fdbt-site/src/data/auroradb.ts:166]: getServicesByNocCodeAndDataSource -> gives the list of active services (whose endDate is either not present and if present should be greater than current date)

getBodsOrTndsServicesByNoc -> gets all the services.

File: repos/fdbt-site/src/utils/index.ts (hunk 3 above)
Comment (NitishAroraKPMG) [reply repos/fdbt-site/src/utils/index.ts:270]: ok

File: repos/fdbt-site/src/data/auroradb.ts (hunk 2 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/data/auroradb.ts:166]: I understand what the DB queries do, but I dont understand why that means why we cant do what i said above?

Comment (RobCattonKPMG) [reply repos/fdbt-site/src/data/auroradb.ts:167]: @NitishAroraKPMG question remains - why do we need to add a new DB query? Why can we not just edit the ones used before?

File: repos/fdbt-site/src/utils/index.ts [hunk 8]
```diff
@@ -261,3 +263,33 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
...
+        if (!uniqueServiceList.includes(serviceName)) {
+            uniqueServices.push(service);
+            uniqueServiceList.push(serviceName);
+        }
+    });
+
+    return uniqueServices;
+};
+
+export const getUniqueFaresServices = (services: MyFaresService[]): MyFaresService[] => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:282]:
```suggestion
export const getUniqueMyFaresServices = (services: MyFaresService[]): MyFaresService[] => {
```

File: repos/fdbt-site/src/utils/index.ts [hunk 9]
```diff
@@ -261,3 +263,33 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
 export const fareTypeIsAllowedToAddACap = (fareType: string): boolean => {
     return !!fareType && ['single', 'return', 'flatFare'].includes(fareType);
 };
+
+export const getUniqueServices = (services: ServiceType[]): ServiceType[] => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:267]: Can we not just have one function which does this, but it just takes this as an argument:

`{ lineId: string, startDate: string, endDate: string | undefined } []`

And then you can just pass the MyFaresService[] and the ServiceType[] to the function, and it wont care that they have additional information?

Comment (NitishAroraKPMG) [reply repos/fdbt-site/src/utils/index.ts:267]: what about the return from the function

File: repos/fdbt-site/src/data/auroradb.ts (hunk 2 above)
Comment (NitishAroraKPMG) [reply repos/fdbt-site/src/data/auroradb.ts:167]: will be deleting the function in next push as it is not referenced anywhere

File: repos/fdbt-site/src/pages/products/services.tsx [hunk 10]
```diff
@@ -188,8 +188,11 @@ export const getServerSideProps = async (ctx: NextPageContextWithSession): Promi
     const noc = getAndValidateNoc(ctx);
     const dataSource = !!getSessionAttribute(ctx.req, MULTI_MODAL_ATTRIBUTE) ? 'tnds' : 'bods';
     const services: MyFaresService[] = await getBodsOrTndsServicesByNoc(noc, dataSource);
+    const unqiueServiceIds = getMyUniqueServices(services);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/products/services.tsx:191]:
```suggestion
    const uniqueServiceIds = getUniqueServiceIds(services);
```

File: repos/fdbt-site/src/utils/index.ts [hunk 11]
```diff
@@ -261,3 +263,50 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
 export const fareTypeIsAllowedToAddACap = (fareType: string): boolean => {
     return !!fareType && ['single', 'return', 'flatFare'].includes(fareType);
 };
+
+export const getMyUniqueServices = (
+    services: { id?: string | number; lineId: string; startDate: string; endDate?: string }[],
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:268]: If we're using id's to store which ones are unique, why is id nullable on the input?

File: repos/fdbt-site/src/utils/index.ts [hunk 12]
```diff
@@ -261,3 +263,50 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
...
+): string[] => {
+    const uniqueServiceList: string[] = [];
+    const uniqueServices: string[] = [];
+
+    services.forEach((service) => {
+        const serviceName = `${service.lineId}#${service.startDate}#${service.endDate ? service.endDate : ''}`;
+        if (!uniqueServiceList.includes(serviceName)) {
+            uniqueServices.push(String(service.id));
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:276]: are id's not strings?

File: repos/fdbt-site/src/pages/service.tsx [hunk 13]
```diff
@@ -108,9 +108,13 @@ export const getServerSideProps = async (ctx: NextPageContextWithSession): Promi
 
     const dataSourceAttribute = getRequiredSessionAttribute(ctx.req, TXC_SOURCE_ATTRIBUTE);
 
-    const services = await getServicesByNocCodeAndDataSource(nocCode, dataSourceAttribute.source);
+    const services: ServiceType[] = await getServicesByNocCodeAndDataSource(nocCode, dataSourceAttribute.source);
+    const unqiueServiceIds = getMyUniqueServices(services);
+    const servicesWithNoDuplicates = services.filter(
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/service.tsx:113]: Is there a reason why we cant move this filter into the function which gets unique services? Why do we have to do it here?

File: repos/fdbt-site/src/utils/index.ts [hunk 14]
```diff
@@ -261,3 +263,50 @@ export const objectKeyMatchesExportNameExactly = (objectKey: string, exportName:
...
+    });
+
+    return uniqueServices;
+};
+
+export const getUniqueServices = (services: ServiceType[]): ServiceType[] => {
+    const uniqueServiceList: string[] = [];
+    const uniqueServices: ServiceType[] = [];
+
+    services.forEach((service) => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/index.ts:288]: I dont think it makes sense for us to have the same code 3 times over, slightly altered.

File: repos/fdbt-site/src/utils/index.ts (hunk 11 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/utils/index.ts:266]: @NitishAroraKPMG needs a response please

File: repos/fdbt-site/tests/utils/index.test.ts [hunk 15]
```diff
@@ -249,4 +251,136 @@ describe('index', () => {
...
+                    lineId: '3h3vb32ik',
+                    startDate: '06/02/2020',
+                    description: 'this bus service is X1',
+                    origin: 'Edinburgh',
+                    serviceCode: 'NW_05_BLAC_X1_1',
+                },
+            ]);
+        });
+
+        const mockFaresServices = [
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/utils/index.test.ts:331]: Can we move this to the top of the describe block please

Comment (RobCattonKPMG) [reply repos/fdbt-site/tests/utils/index.test.ts:331]: And can we also add the type of MyFaresService[] to this variable

File: repos/fdbt-site/tests/utils/index.test.ts [hunk 16]
```diff
@@ -249,4 +251,136 @@ describe('index', () => {
...
+        });
+
+        it.each(unCapFareTypes)('returns false if the fare type is other than %s', (fareType: string) => {
+            const result = fareTypeIsAllowedToAddACap(fareType);
+            expect(result).toBeFalsy();
+        });
+    });
+
+    describe('removeDuplicateServices', () => {
+        const mockServices = [
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/utils/index.test.ts:271]: ServiceType type needs adding to this variable

File: repos/fdbt-site/tests/utils/index.test.ts [hunk 17]
```diff
@@ -249,4 +252,136 @@ describe('index', () => {
...
+            expect(result).toBeTruthy();
+        });
+
+        it.each(unCapFareTypes)('returns false if the fare type is other than %s', (fareType: string) => {
+            const result = fareTypeIsAllowedToAddACap(fareType);
+            expect(result).toBeFalsy();
+        });
+    });
+
+    describe.only('removeDuplicateServices', () => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/utils/index.test.ts:271]:
```suggestion
    describe('removeDuplicateServices', () => {
```

**Review by RobCattonKPMG — APPROVED** [review]
